File: C++/maximum-bitwise-xor-after-rearrangement.cpp
// Time:  O(n)
// Space: O(1)

// greedy
class Solution {
public:
    string maximumXor(string s, string t) {
        vector<int> cnt(2);
        cnt[0] = count(cbegin(t), cend(t), '0');
        cnt[1] = size(t) - cnt[0];
        string result;
        result.reserve(size(s));
        for (const auto& c : s) {
            const auto& x = c - '0';
            if (cnt[x ^ 1]) {
                --cnt[x ^ 1];
                result.push_back('1');
            } else {
                --cnt[x];
                result.push_back('0');
            }
        }
        return result;
    }
};

File: Python/maximum-bitwise-xor-after-rearrangement.py
# Time:  O(n)
# Space: O(1)

# greedy
class Solution(object):
    def maximumXor(self, s, t):
        """
        :type s: str
        :type t: str
        :rtype: str
        """
        cnt = [t.count('0'), t.count('1')]
        result = []
        for i in xrange(len(s)):
            x = ord(s[i])-ord('0')
            if cnt[x^1]:
                cnt[x^1] -= 1
                result.append('1')
            else:
                cnt[x] -= 1
                result.append('0')
        return "".join(result)
